subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java

```diff
@@ -289,11 +291,14 @@ public void outputsRemovedAfterFailureToLoadFromCache() {
         }
 
         @Override
-        public void visitFileInputs(InputFilePropertyVisitor visitor) {
-            for (InputFilePropertySpec inputFileProperty : context.getTaskProperties().getInputFileProperties()) {
+        public void visitInputFileProperties(InputFilePropertyVisitor visitor) {
+            boolean requiresInputChanges = isRequiresInputChanges();
+            boolean usesIncrementalTaskInputs = requiresInputChanges && isUsesIncrementalTaskInputs();
+            ImmutableSortedSet<InputFilePropertySpec> inputFileProperties = context.getTaskProperties().getInputFileProperties();
+            for (InputFilePropertySpec inputFileProperty : inputFileProperties) {
                 Object value = inputFileProperty.getValue();
                 if (value != null) {
-                    visitor.visitInputFileProperty(inputFileProperty.getPropertyName(), value);
+                    visitor.visitInputFileProperty(inputFileProperty.getPropertyName(), value, usesIncrementalTaskInputs || inputFileProperty.isIncremental() || inputFileProperty.isSkipWhenEmpty());
```

Comment:
That condition definitely deserves its own line being assigned to a named variable, probably even its own static function.

Reply:
Done.